### PR TITLE
Add TokenRequestImpl class

### PR DESCRIPTION
**Pull request by dudnyk**
Move internal implementation of TokenRequest to
TokenRequestImpl
Add methods for setup credentials, expiry, body.
Add unit tests

Resolves: OLPEDGE-1265

Signed-off-by: Yevhenii Dudnyk <ext-yevhenii.dudnyk@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 1]
```diff
@@ -41,6 +47,17 @@ class TokenRequest {
   TokenRequest(
       const std::chrono::seconds& expires_in = std::chrono::seconds(0));
 
+  /**
+   * @brief Sets the expiration time.
+   *
+   * @param expires_in (Optional) The number of seconds left before the new
+   * access token expires.
+   *
+   * @return A reference to *this.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:56]: No need to add `@return` to setters. Please remove here and in other setters below.

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 2]
```diff
@@ -21,11 +21,17 @@
...
+#include <memory>
+#include <vector>
 
 #include <olp/authentication/AuthenticationApi.h>
+#include <olp/authentication/AuthenticationCredentials.h>
 
 namespace olp {
 namespace authentication {
+
+using RequestBodyType = std::shared_ptr<const std::vector<std::uint8_t>>;
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:33]: Please add a comment here.

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 3]
```diff
@@ -51,8 +68,43 @@ class TokenRequest {
...
+   * on HERE developer portal.
+   *
+   * @return A reference to *this.
+   */
+  TokenRequest& WithCredentials(AuthenticationCredentials credentials);
+
+  /**
+   * @brief Gets the authentication credentials
+   *
+   * @return reference to `AuthenticationCredentials` instance.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:84]:
```suggestion
   * @return A reference to the `AuthenticationCredentials` instance.
```

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 4]
```diff
@@ -51,8 +68,43 @@ class TokenRequest {
...
+   *
+   * @param credentials Client credentials obtained when registering application
+   * on HERE developer portal.
+   *
+   * @return A reference to *this.
+   */
+  TokenRequest& WithCredentials(AuthenticationCredentials credentials);
+
+  /**
+   * @brief Gets the authentication credentials
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:82]:
```suggestion
   * @brief Gets the authentication credentials.
```

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 5]
```diff
@@ -51,8 +68,43 @@ class TokenRequest {
    */
   std::chrono::seconds GetExpiresIn() const;
 
+  /**
+   * @brief Sets authentication credentials.
+   *
+   * @param credentials Client credentials obtained when registering application
+   * on HERE developer portal.
+   *
+   * @return A reference to *this.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:77]:
```suggestion
   * @brief Sets the authentication credentials.
   *
   * @param credentials The client credentials that you got from the HERE
   * Account as a part of the onboarding or support process on the developer
   * portal.
   *
   * @see `AuthenticationCredentials.h` for more information on the client credentials.
```

**Review by HalynaDumych — CHANGES_REQUESTED** [review]

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 2 above)
Comment (andescu) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:33]: Move to TokenRequest class.

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 6]
```diff
@@ -41,6 +47,17 @@ class TokenRequest {
   TokenRequest(
       const std::chrono::seconds& expires_in = std::chrono::seconds(0));
 
+  /**
+   * @brief Sets the expiration time.
+   *
+   * @param expires_in (Optional) The number of seconds left before the new
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenRequest.h:53]:
```suggestion
   * @param expires_in (Optional) Specifies the seconds the new token shoul be valid for. Maximum allowed value is 86400, means 24 hours. If the value exceeds this value the expires_in value will be ignored by the backend and the default expiration value will be returned.
```

**Review by andescu — COMMENTED** [review]
Although it looks good we should not just add something which is not used. The idea is that this will be together with the small refactoring of TokenProviderPrivate, AutoRefreshingToken & TokenEndpoint, so that the TokenRequest is created by AutoRefreshingToken and not by TokenProviderPrivate, which seems very strange.

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 3 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:84]: done

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 2 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:33]: done

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 1 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:56]: done

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 4 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:82]: done

File: olp-cpp-sdk-authentication/src/TokenRequest.h (hunk 6 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:53]: done

**Pull request closed without merging**